tokio-util/src/lrtd/lrtd.rs

```diff
@@ -0,0 +1,471 @@
+pub trait ThreadStateHandler: Send + Sync {
+    /// Invoked with details of a blocking thread.
+    ///
+    /// # Arguments
+    ///
+    /// * `thread_id` - The ID of the blocking thread.
+    /// * `thread_name` - An optional name of the blocking thread.
+    /// * `backtrace` - The backtrace of the blocking thread.    
+    fn blocking_thread_details(
+        &self,
+        thread_id: libc::pthread_t,
+        thread_name: Option<&str>,
+        backtrace: Backtrace,
+    );
+}
+
+/// A trait for handling actions when blocking is detected.
+///
+/// This trait provides a method for handling the detection of a blocking action.
+pub trait BlockingActionHandler: Send + Sync {
+    /// Called when a blocking action is detected and prior to thread signaling.
+    ///
+    /// # Arguments
+    ///
+    /// * `signal` - The signal used to signal tokio worker threads for state details.
+    /// * `targets` - The list of thread IDs of the tokio runtime worker threads.   /// # Returns
+    ///
+    /// # Returns
+    ///
+    /// Returns `true` if the signaling of the threads is to be executed, false otherwise.
+    ///
+    fn blocking_detected(&self, signal: Signal, targets: &Vec<libc::pthread_t>) -> bool;
+}
+
+struct StdErrBlockingActionHandler;
+
+/// BlockingActionHandler implementation that writes blocker details to standard error.
+impl BlockingActionHandler for StdErrBlockingActionHandler {
+    fn blocking_detected(&self, signal: Signal, targets: &Vec<libc::pthread_t>) -> bool {
+        eprintln!(
+            "Detected worker blocking, signaling {} worker threads: {:?}",
+            signal, targets
+        );
+        true
+    }
+}
+
+struct StdErrThreadStateHandler;
+
+/// ThreadStateHandler implementation that writes details to standard error.
+impl ThreadStateHandler for StdErrThreadStateHandler {
+    fn blocking_thread_details(
+        &self,
+        thread_id: libc::pthread_t,
+        thread_name: Option<&str>,
+        backtrace: Backtrace,
+    ) {
+        let name = thread_name
+            .map(|n| format!(" for thread \"{}\"", n))
+            .unwrap_or("".to_owned());
+        eprintln!("Stack trace{}:{}\n{}", name, thread_id, backtrace);
+    }
+}
+
+#[derive(Debug)]
+struct WorkerSet {
+    inner: Mutex<HashSet<libc::pthread_t>>,
+}
+
+impl WorkerSet {
+    fn new() -> Self {
+        WorkerSet {
+            inner: Mutex::new(HashSet::new()),
+        }
+    }
+
+    fn add(&self, pid: libc::pthread_t) {
+        let mut set = self.inner.lock().unwrap();
+        set.insert(pid);
+    }
+
+    fn remove(&self, pid: libc::pthread_t) {
+        let mut set = self.inner.lock().unwrap();
+        set.remove(&pid);
+    }
+
+    fn get_all(&self) -> Vec<libc::pthread_t> {
+        let set = self.inner.lock().unwrap();
+        set.iter().cloned().collect()
+    }
+}
+
+extern "C" fn signal_handler(_: i32) {
+    let backtrace = Backtrace::force_capture();
```

Comment:
This is not async-signal safe.

Reply:
though so, any ideas on a better approach?

Reply:
Separated thread state dump out of LRTD.